_source/_posts/2021-12-03-antipatterns-secrets-java.adoc

```diff
@@ -0,0 +1,149 @@
+
+If you find a secret in your source, remove it and revoke it. It's best if secrets never get checked in the first place, watch out for them in code reviews and scan your code bases to ensure this doesn't happen: https://www.gitguardian.com/[GitGuardian] and https://github.com/awslabs/git-secrets[git-secrets].
+
+Secrets often enter source trees by accident, usually because you temporarily typed one into a config file. When secrets are mixed in with other application properties, it makes them harder to spot or simply forgotten about. Keep secrets out of your source entirely by using an https://12factor.net/config[external configuration] provider like Hashicorp Vault, using environmental variables], or by keeping secrets in a separate file outside of your source tree (or one that is explicitly ignored in your `.gitignore` file.)
+
+TIP: Check out the https://github.com/paulschwarz/spring-dotenv[spring-dotenv] to use `.env` files with your Spring projects.
+
+
+== 2. Confusing encoding with encryption
+
+Many folks are confused about the difference between encoding and encryption; they are very distinct operations and cannot be interchanged.
+
+=== Encoding
+
+Data that is _encoded_ is transformed into another format, usually to make transferring between systems possible. An encoded value may or may not be human-readable, but it needs to be treated the same way as the original input source.  The following examples show a string encoded in two common encoding schemes URL encoding.
+
+URL Encoded:
+
+[source,java]
+----
+String text = "Can you still read this text?";
+String encoded = URLEncoder.encode(text, UTF_8);
+// encoded: Can+you+still+read+this+text%3F
+String decoded = URLDecoder.decode(encoded, UTF_8);
+assertThat(decoded, equalTo(decoded));
+----
+
+Base64 encoded (not human-readable):
+
+[source,java]
+----
+String text = "Can you still read this text?";
+String encoded = Base64.getEncoder().encodeToString(text.getBytes(UTF_8));
+// encoded: Q2FuIHlvdSBzdGlsbCByZWFkIHRoaXMgdGV4dD8=
+String decoded = new String(Base64.getDecoder().decode(encoded.getBytes(UTF_8)));
+assertThat(decoded, equalTo(decoded));
+----
+
+IMPORTANT: In both cases, the encoded data can easily be reversed, and the encoded string must be treaded the same way as plain text.
+
+=== Encryption
+
+Data that is encrypted can only be read by authorized parties (anyone with the correct cryptographic key). When data is encrypted, the resulting value is indistinguishable from random data. Because of this, encrypted data is often also encoded to make it easier to transport. For example, in the case of JSON Web Encryption (JWE), the JSON payload is encrypted and then bas64 encoded string.
```

Comment:
```suggestion
In contrast, data that's encrypted can only be read by authorized parties (anyone with the correct cryptographic key). When data is encrypted, the resulting value is indistinguishable from random data. Because of this, encrypted data is often also encoded to make it easier to transport. For example, in the case of JSON Web Encryption (JWE), the JSON payload is encrypted and is then transformed to a base64-encoded string.
```
typo +edits